src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,886 @@
+from flint.pyflint cimport global_random_state
+from flint.types.fmpz cimport fmpz, any_as_fmpz
+from flint.types.nmod cimport nmod
+from flint.types.fmpz_mod cimport fmpz_mod
+from flint.types.fmpz_poly cimport fmpz_poly, any_as_fmpz_poly, fmpz_poly_set_list
+from flint.types.fmpz_mod_poly cimport fmpz_mod_poly, fmpz_mod_poly_ctx
+from flint.types.nmod_poly cimport nmod_poly
+from flint.utils.typecheck cimport typecheck
+
+cdef class fq_default_ctx:
+    r"""
+    Context object for creating :class:`~.fq_default`.
+
+    Finite fields can be initialized in one of two possible ways. The
+    first is by providing characteristic and degree:
+
+        >>> fq_default_ctx(5, 2, 'y', fq_type=1)
+        fq_default_ctx(5, 2, 'y', x^2 + 4*x + 2, 'FQ_ZECH')
+
+    The second is by giving an irreducible polynomial of type
+    :class:`~.nmod_poly` or :class:`~.fmpz_mod_poly`:
+
+        >>> from flint import fmpz_mod_poly_ctx
+        >>> mod = fmpz_mod_poly_ctx(11)([1,0,1])
+        >>> fq_default_ctx(modulus=mod, fq_type=2)
+        fq_default_ctx(11, 2, 'x', x^2 + 1, 'FQ_NMOD')
+
+    For more details, see the documentation of :method:`~.set_from_order`
+    and :method:`~.set_from_modulus`.
+    """
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        if self._initialized:
+            fq_default_ctx_clear(self.val)
+
+    @staticmethod
+    def _parse_input_fq_type(fq_type):
+        # Allow the type to be denoted by strings or integers
+        FQ_TYPES = {
+            "FQ_ZECH" : 1,
+            "FQ_NMOD" : 2,
```

Comment:
This dictionary is already provided by the enum type:
```python
In [35]: fq_default_type._member_map_
Out[35]: 
{'DEFAULT': <fq_default_type.DEFAULT: 0>,
 'FQ_ZECH': <fq_default_type.FQ_ZECH: 1>,
 'FQ_NMOD': <fq_default_type.FQ_NMOD: 2>,
 'FQ': <fq_default_type.FQ: 3>,
 'NMOD': <fq_default_type.NMOD: 4>,
 'FMPZ_MOD': <fq_default_type.FMPZ_MOD: 5>}
```

Reply:
I guess this function can just be:
```python
try:
    return fq_default_type._member_map_[fq_type]
except KeyError:
    pass

try:
    return fq_default_type._value2member_map_[fq_type]
except KeyError:
    pass

raise ValueError(f"fq_type should be one of {fq_default_type._member_names_}")

Reply:
I'm getting an error that `fq_default_type` doesn't have a member map... I'll try and figure out why

Reply:
I have it available in python:

```py
In [1]: from flint import fq_default_ctx, fq_default_type

In [2]: fq_default_type
Out[2]: <flag 'fq_default_type'>

In [3]: fq_default_type._member_map_
Out[3]: 
{'DEFAULT': <fq_default_type.DEFAULT: 0>,
 'FQ_ZECH': <fq_default_type.FQ_ZECH: 1>,
 'FQ_NMOD': <fq_default_type.FQ_NMOD: 2>,
 'FQ': <fq_default_type.FQ: 3>,
 'NMOD': <fq_default_type.NMOD: 4>,
 'FMPZ_MOD': <fq_default_type.FMPZ_MOD: 5>}

In [4]: type(fq_default_type)
Out[4]: enum.EnumType
```

But for compilation of the cython I get:

```
      Error compiling Cython file:
      ------------------------------------------------------------
      ...
              return fq_type
      
          @staticmethod
          def _parse_input_var(var):
              try:
                  return fq_default_type._member_map_[fq_type]
                                        ^
      ------------------------------------------------------------
      
      /Users/Jack/Documents/GitHub/python-flint/src/flint/types/fq_default.pyx:65:34: _member_map_ not a known value of fq_default_type
```

Reply:
I guess that the C-level enum is different in the cpdef enum.

Maybe using the enum type for validation isn't quite a good fit and we just need to have a dict somewhere to map values from one to the other.

Either way we need there to be a single place where the list of values is stored rather than several different dicts spread around.